torax/sources/ec_current_source.py

```diff
@@ -0,0 +1,123 @@
+    ec_power_density: runtime_params_lib.TimeInterpolated
+
+    def make_provider(self, torax_mesh: geometry.Grid1D | None = None):
+        if torax_mesh is None:
+            raise ValueError("torax_mesh is required for RuntimeParams.make_provider.")
+        return RuntimeParamsProvider(**self.get_provider_kwargs(torax_mesh))
+
+
+@chex.dataclass
+class RuntimeParamsProvider(runtime_params_lib.RuntimeParamsProvider):
+    """Provides runtime parameters for the electron-cyclotron current source for a given time and geometry."""
+
+    runtime_params_config: RuntimeParams
+    ec_power_density: interpolated_param.InterpolatedVarTimeRho
+    global_efficiency: interpolated_param.InterpolatedVarSingleAxis
+
+    def build_dynamic_params(
+        self,
+        t: chex.Numeric,
+    ) -> DynamicRuntimeParams:
+        return DynamicRuntimeParams(**self.get_dynamic_params_kwargs(t))
+
+
+@chex.dataclass(frozen=True)
+class DynamicRuntimeParams(runtime_params_lib.DynamicRuntimeParams):
+    """Runtime parameters for the electron-cyclotron current source for a given time and geometry."""
+
+    ec_power_density: array_typing.ArrayFloat
+    global_efficiency: array_typing.ScalarFloat
+
+
+def _calc_eccd_current(
+    static_runtime_params_slice: runtime_params_slice.StaticRuntimeParamsSlice,
+    dynamic_runtime_params_slice: runtime_params_slice.DynamicRuntimeParamsSlice,
+    dynamic_source_runtime_params: runtime_params_lib.DynamicRuntimeParams,
+    geo: geometry.Geometry,
+    core_profiles: state.CoreProfiles,
+) -> jax.Array:
+    """Model function for the ECRH current source.
+
+    Calculated as:
+
+    .. math::
+      j_{EC} = \\frac{\\epsilon_0^2}{q_e^3} \\eta_{cd} \\frac{q_{EC} T_e}{n_e R_{maj}}
+
+    where:
+    - :math:`j_{EC}` is the flux-surface averaged EC current drive profile in A/m^2,
+    - :math:`\\epsilon_0` is the permittivity of free space in SI units,
+    - :math:`q_e` is the elementary charge in C,
+    - :math:`\\eta_{cd}` is the dimensionless global current drive efficiency,
+    - :math:`q_{EC}` is the EC power density in W/m^2,
+    - :math:`T_e` is the electron temperature in J,
+    - :math:`n_e` is the electron density in m^-3,
+    - :math:`R_{maj}` is the major radius in m.
+    """
+    assert isinstance(dynamic_source_runtime_params, DynamicRuntimeParams)
+
+    # Sources:
+    # Electron cyclotron current drive efficiency in general tokamak geometry
+    #   Lin-Liu, Chan, and Prater, 2003
+    #   https://doi.org/10.1063/1.1610472
+    # Flat-top plasma operational space of the STEP power plant
+    #   Tholerus et al., 2024
+    #   https://doi.org/10.1088/1741-4326/ad6ea2
+
+    ne_face_in_m3 = (
+        core_profiles.ne.face_value() * dynamic_runtime_params_slice.numerics.nref
+    )
+    Te_face_in_J = core_profiles.temp_el.face_value() * CONSTANTS.keV2J
+
+    # Flux-surface averaged j profile, <j_ec> [A/m^2]
+    return (
+        CONSTANTS.epsilon0**2 # [m^-3 kg^-1 s^4 A^2]
```

Comment:
This formula will struggle with fp32 underflows. Currently we use fp64 but reverting to fp32 is on the roadmap.

Better to do a logarithm version of the formula and then exponentiate it. See e.g. https://github.com/google-deepmind/torax/blob/53390ca796bf67b2f51841e5e803301bfcacd7bd/torax/physics.py#L81